packages/sandcastle/src/Gallery.tsx

```diff
@@ -15,6 +16,140 @@ export type GalleryItem = {
+  const [pagefindLoaded, setPagefindLoaded] = useState(false);
+  const [availableFilters, setAvailableFilters] =
+    useState<PagefindFilterCounts>({ tags: {} });
+
+  const [searchTerm, setSearchTerm] = useState("");
+  const [currentTag, setCurrentTag] = useState<string | "All">("All");
+
+  const requestId = useRef<number>(0);
+
+  useEffect(() => {
+    async function doSearch() {
+      if (!pagefind.current) {
+        return;
+      }
+      const searchString = searchTerm;
+      if (!searchString || searchString === "") {
+        setSearchResults(null);
+        return;
+      }
+      const thisRequestId = ++requestId.current;
+
+      const search = await pagefind.current.search(searchString, {
+        filters: {
+          tags: currentTag === "All" ? undefined : currentTag,
+        },
+      });
+
+      const resultDataPromises = await Promise.allSettled(
+        search.results.map((result) => {
+          return result.data();
+        }),
+      );
+      const resultData = resultDataPromises
+        .map((result) => {
+          return result.status === "fulfilled" ? result.value : undefined;
+        })
+        .filter((result) => !!result);
+
+      if (thisRequestId === requestId.current) {
+        setSearchResults(resultData);
+      }
+    }
+
+    doSearch();
+  }, [searchTerm, currentTag, setSearchResults]);
+
+  useEffect(() => {
+    async function loadPageFind() {
+      const pagefindImport: Pagefind = await import(
+        // vite doesn't like the dynamic import
+        /* @vite-ignore */
+        `${GALLERY_BASE}/pagefind/pagefind.js`
+      );
+
+      pagefindImport.init();
+      await pagefindImport.options({
+        baseUrl: getBaseUrl(),
+      });
+      pagefind.current = pagefindImport;
+
+      const filters = await pagefind.current.filters();
+      setAvailableFilters(filters);
+
+      setPagefindLoaded(true);
+    }
+    if (isFirstLoad.current) {
+      // This prevents double activation when using StrictMode
+      isFirstLoad.current = false;
+      loadPageFind();
+    }
+  }, []);
+
+  const debounceTimeout = useRef<NodeJS.Timeout>(null);
+
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  return (
+    <>
+      {!!searchTerm && (
+        <Button
+          onClick={() => {
+            setSearchTerm("");
+            if (inputRef.current) {
+              inputRef.current.value = "";
+            }
+          }}
+        >
+          Clear
+        </Button>
+      )}
+      <Input
+        ref={inputRef}
+        disabled={!pagefindLoaded}
+        onChange={(e) => {
+          if (debounceTimeout.current) {
+            clearInterval(debounceTimeout.current);
+          }
+          const newSearchTerm = e.target.value;
+          pagefind.current?.preload(newSearchTerm);
+          debounceTimeout.current = setTimeout(() => {
+            setSearchTerm(newSearchTerm);
+          }, 300);
+        }}
+        placeholder="Filter demos..."
```

Comment:
```suggestion
        placeholder="Search gallery"
```

1. Let's use "Search" since all users seem to be calling it that
2. Let's use "gallery" to be specific about _where_ we're searching
3. Doing a quick comparison, I don't see a lot of `...` appended to search placeholders.